### ISSUE TITLE
Electron branch?

### ISSUE THREAD

**Issue by colinskow** [AngularClass/angular2-webpack-starter#1352]
I am going to be creating a version of this repo specifically for building [Electron](http://electron.atom.io) apps. Electron is quite heavy so this would be inappropriate for the main branch.

Main changes will be:
- Inclusion of Electron dependencies
- The addition of a separate build for the Electron main process
- Karma tests will run inside Electron
- E2E tests will be redone to use [Spectron](http://electron.atom.io/spectron/) instead of Protractor

The current Angular 2 build process will remain practically unchanged, so it should be easy to maintain those files in sync with this repo.

My question: is AngularClass interested in maintaining this in a separate branch, or should I build it as my own fork? Once it is built, I could definitely use help maintaining it as technology evolves.

**Comment by d3viant0ne** [comment]
We have historically not been very consistent in the maintenance of feature branches and I have always been partial to the approach @mgechev took where major features were maintained in forks with a link to them added to the readme of the base.

Keeps the issues & pull requests focused a bit better imo.

**Comment by colinskow** [comment]
OK, I'll go ahead and make this a complete fork then.

**Comment by d3viant0ne** [comment]
Don't take my opinion as law

**Comment by katallaxie** [comment]
@colinskow for the sake of maintainability I would fork the repo. I also have my very own for Electron support. I think this needs special care, the same way as you want to support NativeScript.

**Comment by colinskow** [comment]
@katallaxie would you please give me a link to your Electron repo, this could save me a lot of work if you've done it.

**Comment by d3viant0ne** [comment]
https://github.com/katallaxie/angular2-electron-starter

**Comment by JonnyBGod** [comment]
I am also maintaining a starter based on AngularClass's with electron support and at some point NativeScript.

https://github.com/JonnyBGod/angular2-webpack-advance-starter

**Comment by colinskow** [comment]
Thank you very much @JonnyBGod. You've given me some great ideas.

I'm currently working on setting up double watch builds, one for Electron and one for the Renderer. Whenever an Electron dependency is updated it should restart Electron (via electron-connect). And then I will connect the Renderer process to the Hot Module Reloader via a [custom server script](https://github.com/chentsulin/electron-react-boilerplate/blob/master/server.js).

I found a [bug](https://github.com/s-panferov/awesome-typescript-loader/issues/351) in `awesome-typescript-loader` that causes Webpack to watch every single .ts file in the project, no matter what. So I'm working on getting the main and renderer builds to refresh independently only when their dependencies are updated. (Right now I've resorted to monkey-patching my installation of ATL.)

The other project I am working on is replacing Protractor with [Spectron](http://electron.atom.io/spectron/) since Protractor doesn't work well with Electron.

Maybe we can work together or at least rip code off each other ;-)

**Comment by JonnyBGod** [comment]
@colinskow Yes I am glad to help with it.

Let me have your repo link and I will keep track of it and help when possible.

You can use my code where you need, no problem at all. Also, if ou would like to implement your improvements in my repo that would be awesome.

**Comment by colinskow** [comment]
I will post it here as soon as I've implemented the basic functionality.

**Comment by colinskow** [comment]
@JonnyBGod it's ready!!!
https://github.com/colinskow/angular-electron-dream-starter

**Comment by JonnyBGod** [comment]
@colinskow Cool that was fast. Will test it soon.

Quick remarks from my very quick look at it:

1. I think electron.main should be on a dedicated folder in root. This is because you potentially want to extend electron main capabilities and have multiple files for it.

2. I don't think NGRX fits this starter. That is why I created an advanced starter. NGRX is just too complex for most dev starting with angular.

Would you be willing to implement your electron improvements in my [advanced starter](https://github.com/JonnyBGod/angular2-webpack-advance-starter)?

**Comment by colinskow** [comment]
@JonnyBGod thank you for the feedback.

1. I created a separate `electron` folder inside `src` to place sub-modules for the main process. It felt awkward having them outside of the `src` tree. Ideally I would use a different folder structure, but I want to keep everything the same so I can easily pull updates from this repo.

2. Most devs who are just starting with Angular2 should be using Angular-CLI. This repo is really for those of us who aren't satisfied with CLI and are obsessed with customizing the build. Redux/ngrx is the absolute best way to manage state and scaling without it is a headache. I included a very simple implementation (one reducer, one action). It wanted to show how to hook it up to HMR, unit tests, and the Redux DevTools extension since that is tricky.

I just setup several DevTools extensions which will automatically install themselves in development mode.

Feel free to pull what you want from my code. I'm using specialized Electron code, so my fork no longer supports browser builds.
